docs/proposals/config-management-plugin-v2.md

```diff
@@ -0,0 +1,304 @@
+
+#### Discovery (Auto-selection of Tool)
+
+- The plugin discovery will run in the main repo-server container.
+- Argo CD repo-server lists the shared plugins directory and runs `discover` command from the specification file, 
+whichever plugin provides a positive response first will be selected.
+- In case of multiple plugins, repo-server will decide based on the plugin priority list set by argocd-server. 
+This priority list can be set in the Argo CD config map. 
+
+#### Versioning
+There will be one image per version. Hence, for two different versions users will have to configure two different sidecars.
+
+### Security Considerations
+
+The use of the plugin as sidecars separate from the repo-server is already a security improvement over the current v1.8 
+config management plugin mechanism, since the plugin tooling will no longer have access to the files of the argocd-repo-server image. 
+However additional improvements can be made to increase security.
+
+Currently, nothing prevents a command from traversing outside of the git repository, and reading/writing files outside 
+the repository path, including other git repositories which were cloned next to the working directory of the generate command. 
+Although the generate commands which are executed are determined by an operator, they are still affected by end user input. 
+For example, an  end-user might try to exploit the reposerver by specifying `HEAD && rm -rf /` as a git targetRevision, 
+which they might expect to execute as:
+
+`git clone HEAD && rm -rf /`
+
+(Note that this is just used as an example and the repo-server is not actually exploitable in this way)
+
+### Risks and Mitigations
+
+One issue is that currently when repositories are cloned, the repo is cloned using the same UID of the repo-server user, 
+and so all repository files are created using that UID. This means that a command which executes in the git repository path, 
+could traverse upwards and see/write files which are outside of the repository tree.
+
+One proposal to prevent out-of-tree access to files, is that each git repository could be cloned with unique UIDs, 
+different from the repo-server’s UID. When the cmp-server executes the tooling command to generate manifests, 
+the command could be executed using the UID of the git repository files. e.g.:
+
+```
+cmd := exec.Command(command, args...)
+cmd.SysProcAttr = &syscall.SysProcAttr{}
+cmd.SysProcAttr.Credential = &syscall.Credential{Uid: uid, Gid: gid}
+```
+
+This would ensure that the command could not read or write anything out-of-tree from the repository directory.
+
+### Upgrade / Downgrade Strategy
+
+The argocd-repo-server manifest will change in order to populate the argocd-cmp-server binary inside the plugin container 
+via an init container, which will pre-populate a volume shared between plugins and the repo-server.
+
+```bash
+# An init container will copy the argocd static binary into the shared volume
```

Comment:
Is this being done to take care of the existing plug-ins like Helm & Kustomize?

Reply:
This will be ultimate goal, but out of scope for this proposal.